examples/reporting/parallel_report_download.py

```diff
@@ -0,0 +1,191 @@
+                failures.append(res[1])
+
+        # Output results.
+        print(
+            f'Total successful results: {len(successes)}\n'
+            f'Total failed results: {len(failures)}\n'
+        )
+
+        print("Successes:") if len(successes) else None
+        for success in successes:
+            # success["results"] represents an array of result strings for one
+            # customer ID / query combination.
+            result_str = "\n".join(success["results"])
+            print(result_str)
+
+        print("Failures:") if len(failures) else None
+        for failure in failures:
+            ex = failure["exception"]
+            print(
+                f'Request with ID "{ex.request_id}" failed with status '
+                f'"{ex.error.code().name}" for customer_id '
+                f'{failure["customer_id"]} and query "{failure["query"]}" and '
+                'includes the following errors:'
+            )
+            for error in ex.failure.errors:
+                print(f'\tError with message "{error.message}".')
+                if error.location:
+                    for (
+                        field_path_element
+                    ) in error.location.field_path_elements:
+                        print(f'\t\tOn field: {field_path_element.field_name}')
+
+
+def _issue_search_request(client, customer_id, query):
+    """Issues a search request using streaming.
+
+    Retries if a GoogleAdsException is caught, until MAX_RETRIES is reached.
+
+    Args:
+        client: an initialized GoogleAdsClient instance.
+        customer_id: a client customer ID str.
+        query: a GAQL query str.
+    """
+    ga_service = client.get_service("GoogleAdsService", version="v6")
+    retry_count = 0
+    # Retry until we've reached MAX_RETRIES or have successfully received a
+    # response.
+    while True:
+        try:
+            response = ga_service.search_stream(customer_id, query)
+            # Returning a list of GoogleAdsRows will result in a
+            # PicklingError, so instead we put the GoogleAdsRow data
+            # into a list of str results and return that.
+            result_strings = []
+            for batch in response:
+                for row in batch.results:
+                    ad_group_id = (
+                        f'Ad Group ID {row.ad_group.id} in '
+                        if 'ad_group.id' in query
+                        else ''
+                    )
+                    result_string = (
+                        f'{ad_group_id}''
+                        f'Campaign ID {row.campaign.id} '
+                        f'had {row.metrics.impressions} impressions '
+                        f'and {row.metrics.clicks} clicks.'
+                    )
+                    result_strings.append(result_string)
+            return (True, {"results": result_strings})
+        except GoogleAdsException as ex:
+            if retry_count < MAX_RETRIES:
+                retry_count += 1
```

Comment:
Strictly speaking, we should retry only if the error is of a retriable nature -- e.g. database error or concurrent modification or such. Perhaps we can leave a comment here to say that this code example retries on all errors, but in practice, you'd want it to limit to a handful of errors that makes sense for the developer.

Reply:
added a comment